cmd/datamon/cmd/cli_fuse_test.go

```diff
@@ -3,166 +3,331 @@
 package cmd
 
 import (
+	"bufio"
+	"io"
 	"io/ioutil"
+	"math/rand"
 	"os"
 	"os/exec"
 	"path/filepath"
+	"regexp"
+	"strconv"
 	"strings"
+	"sync"
 	"testing"
 	"time"
 
+	"github.com/oneconcern/datamon/pkg/dlogger"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
 )
 
-func testBundleMount(t *testing.T, testType string) {
+var (
+	buildBinaryonce sync.Once
+	testBinary      string
+)
+
+func testLogger(t testing.TB) *zap.Logger {
+	// a logger for the test sequence itself
+	return dlogger.MustGetLogger("debug").With(zap.String("test", t.Name()))
+}
+func testCommand(t testing.TB, target string, args ...string) (*exec.Cmd, io.ReadCloser) {
+	// NOTE: this starts a process, not a goroutine like for other tests
+	// That is primarily because we want to send a signal to the process
+	// in order to unmount.
+	cmd := exec.Command(target, args...)
+	cmd.Env = append(cmd.Env, "DATAMON_CONTEXT="+os.Getenv("DATAMON_CONTEXT")) // TODO(fred): not sure this is used
+	cmd.Env = append(cmd.Env, "DATAMON_GLOBAL_CONFIG="+os.Getenv("DATAMON_GLOBAL_CONFIG"))
+	cmd.Env = append(cmd.Env, "HOME="+os.Getenv("HOME"))
+	cmd.Env = append(cmd.Env, "PATH="+os.Getenv("PATH"))
+	cmd.Env = append(cmd.Env, "GOOGLE_APPLICATION_CREDENTIALS="+os.Getenv("GOOGLE_APPLICATION_CREDENTIALS"))
+	pipeOut, err := cmd.StdoutPipe()
+	require.NoError(t, err)
+	pipeErr, err := cmd.StderrPipe()
+	require.NoError(t, err)
+
+	// combine stdout & stderr, tee this output to os.Stdout and return pipe reader for output scanning
+	pipeR, pipeW := io.Pipe()
+	c := io.MultiWriter(os.Stdout, pipeW)
+
+	go func() {
+		_, _ = io.Copy(c, pipeOut)
```

Comment:
i'm trying to understand golang stdlib api here:

why use `MultiWriter` to fan out but two `Copy` calls instead of a `MultiReader` and one `Copy`?  are the two equivalent?

Reply:
this part has been a pain, really.
The way output is piped from the forked process looks like a bit dodgy because of the way the `cmd` implementation is copying the output... 

I've tried many combinations along the lines of what you are describing (TeeReader, MultiWriter, etc...). I admit there is still room for improvement here, and I don't have the answers to all the _whys_ in this section.

The main issue here was to be able to catch the _normally unbuffered_ stdout & stderr _before_ the `cmd.Wait()`. It turned out that an explicit copy to a pipe was needed to accomplish that. Any more elegant combination I've tried did actually fail to capture the output progress.

Reply:
follow-up note: open discussion thread

Reply:
ok, so what i'm taking away from this is that there's a buffer on stdout and stderr somewhere, and we need to flush it before `.Wait()`, and this ensures that the buffer gets flushed, whereas a `Copy` on `MultiReader` doesn't.